File: PE_0643/PE_0643.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Mon Jan 20 05:39:22 2020

PE_0643

2-Friendly numbers

968274154

@author: mbh
"""

import time

def p643(n=10**11,modval=10**9+7):
    t0=time.perf_counter()
    ans=0
    while n>1:
        n=n//2
        increment=totientSum(n)
        ans=(ans+increment-1)
    #     print(count-1,increment,increment>2**63-1)
    
    # print (ans%modval)
    print (ans)
    print(time.perf_counter()-t0)



#implements stack exchange 'Andy' 
#http://math.stackexchange.com/questions/316376/how-to-calculate-these-totient-summation-sums-efficiently
# @nb.njit
def R2(N,X2={1:1}):
    if N==1:
        return 0
    try:
        return X2[N]
    except KeyError:
        fsum = F2(N)
        m=2
        while 1:
            x = N//m
            nxt = N//x
            if nxt >= N:
                result=fsum -(N-m+1)*R2(N//m,X2)
                X2[N]=result
                return result
            fsum -= (nxt-m+1) * R2(N//m,X2)
            m = nxt+1


def F2(n):
    return n*(n-1)//2

#returns sum of totients of x<=n
#wrapper for R2
#sum of totient(x) for x<=n
def totientSum(n):
    return R2(n)+1

def F2mod(n,modval):
    return ((n%modval*(n-1)%modval)*inverse(2,modval))%modval

    


#from Wikipedia
def inverse(a, n):
    """returns multiplicative inverse of a mod n. a and n must be-co-prime"""
    t1,t2=0,1    
    r1,r2=n,a    
    while r2!=0:
        q = r1 // r2
        t1, t2 = t2, t1 - q * t2
        r1, r2 = r2, r1 - q * r2
    if t1 < 0:
        t1 +=n
    return t1 

File: PE_0643/PE_0643.jl
# PE 0643
# January 18 2020
# Michael Hunt

# 2-Friendly

# n=10^10: 292126961
# n=10^11: 968274154

# I eventually (it is almost always eventually with me) realised that what we want is
#
# $$ \sum_{d=2^k: d | n}\Phi(d)=\sum_d\sum_{k=1}^d\phi(k)$$
# where $\phi(n)$ is the totient function.

using Memoize
using BenchmarkTools

const M = 10^9+7
pow10 = 11

# As used in PE512 -
#implements stack exchange 'Andy'
#http://math.stackexchange.com/questions/316376/how-to-calculate-these-totient-summation-sums-efficiently
@memoize function R2(n)::Int64
        #R2(n)+1 is the totient sum: Φ(n) = Σ ϕ(k) for k ∈ 1:n
    if n==1
        return 0
    else
        fsum = mod(mod(mod(n,M)*mod((n-1),M),M)*invmod
        (2,M),M)
        m=2
        while true
            x = div(n,m)
            nxt = div(n,x)
            if(nxt >= n)
                return mod(mod(fsum,M) - mod(mod(n-m+1,M) * mod(R2(x),M),M),M)
            end
            fsum = mod(mod(fsum,M) - mod(mod(nxt-m+1,M) * mod(R2(x),M),M),M)
            m = nxt+1
        end
    end
end

p643(n)=mod(reduce(+,R2(div(n,2^k)) for k in 1:ndigits(n; base=2)-1; init=0),M)

@show pow10
@show @benchmark p643(10^pow10)
@show p643(10^pow10)
